File: sdk/dotnet/NoxMem.Client/Errors/NoxMemApiException.cs
namespace NoxMem.Client.Errors;

/// <summary>
/// Thrown when the memoria-nox server returns a non-2xx HTTP response.
/// </summary>
public class NoxMemApiException : HttpRequestException
{
    /// <summary>HTTP status code returned by the server.</summary>
    public int StatusCode { get; }

    /// <summary>Raw response body from the server.</summary>
    public string ResponseBody { get; }

    /// <summary>Request URL that triggered the error.</summary>
    public string Url { get; }

    public NoxMemApiException(int statusCode, string responseBody, string url)
        : base($"NoxMem API error {statusCode} on {url}: {responseBody}")
    {
        StatusCode = statusCode;
        ResponseBody = responseBody;
        Url = url;
    }

    /// <summary>
    /// True when the server returned 503 with
    /// <c>{"error":"feature disabled","env_var":"NOX_*"}</c>.
    /// </summary>
    public bool IsFeatureDisabled =>
        StatusCode == 503 && ResponseBody.Contains("\"feature disabled\"");

    /// <summary>True when the server returned 401 Unauthorized.</summary>
    public bool IsUnauthorized => StatusCode == 401;
}

File: sdk/java/src/main/java/com/noxmem/client/error/NoxMemApiException.java
package com.noxmem.client.error;

/**
 * Thrown when the memoria-nox server returns a non-2xx HTTP status.
 */
public class NoxMemApiException extends RuntimeException {

    private final int statusCode;
    private final String responseBody;
    private final String url;

    public NoxMemApiException(int statusCode, String responseBody, String url) {
        super(String.format("NoxMem API error %d on %s: %s", statusCode, url, responseBody));
        this.statusCode = statusCode;
        this.responseBody = responseBody;
        this.url = url;
    }

    public int getStatusCode() { return statusCode; }
    public String getResponseBody() { return responseBody; }
    public String getUrl() { return url; }

    /** True when the server returned 503 with {"error":"feature disabled",...}. */
    public boolean isFeatureDisabled() {
        return statusCode == 503 && responseBody.contains("\"feature disabled\"");
    }

    /** True when the server returned 401 Unauthorized. */
    public boolean isUnauthorized() { return statusCode == 401; }
}
